File: Java/ExtensibleCommands/src/main/java/org/extensiblecommands/ExtensibleCommandsCore.java
package org.extensiblecommands;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

/**
 * Static class representing an instance of Extensible Commands.
 * Implements static properties and methods that are applicable to generic Command behavior.
 */
public class ExtensibleCommandsCore {
    /**
     * Enumeration for Log levels
     */
    public enum LogLevel {
        Info, Error
    }

    /**
     * Set Logger object
     * @param externalLogger    Logger object implementing Log interface
     */
    public static void setLogger(Log externalLogger) {
        logger = externalLogger;
    }

    /**
     * @return              Logger object implementing Log interface
     */
    public static Log getLogger() {
        return logger;
    }

    /**
     * @return              Timestamp string
     */
    public static String getTimeStamp() {
        return DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm:ss:SSS").format(LocalDateTime.now());
    }

    /**
     * Set Logging Enabled flag
     * @param isEnabled     Is logging of command events enabled?
     */
    public static void setIsLoggingEnabled(boolean isEnabled) {
        isLoggingEnabled = isEnabled;
    }

    /**
     * @return              Is logging of command events enabled?
     */
    public static boolean getIsLoggingEnabled() {
        return isLoggingEnabled;
    }

    /**
     * Log informational or error message
     * @param logLevel      Log level
     * @param message       Message to be logged
     */
    public static void log(LogLevel logLevel, String message) {
        if (getLogger() == null) return;

        // Always log errors regardless of the logging enabled flag
        if (getIsLoggingEnabled() || logLevel == LogLevel.Error)
            logger.log(getTimeStamp(), logLevel.toString(), message);
    }

    //----------------------------------------------------------------------------------------------------------------------

    private static Log logger;
    private static boolean isLoggingEnabled = true;
}

File: CSharp/ExtensibleCommands/ExtensibleCommands/ExtensibleCommandsCore.cs
using System;

namespace ExtensibleCommands
{
    /// <summary> 
    /// Implements static properties and methods that are applicable to generic Command behavior.
    /// </summary>
    public static class ExtensibleCommandsCore
    {
        /// <summary> Enumeration for Log levels </summary>
        public enum LogLevel
        {
            Info,
            Error
        };

        /// <summary> Externally supplied Logger object implementing ILog interface </summary>
        public static ILog Logger { get; set; }

        /// <summary> Timestamp string </summary>
        public static string TimeStamp
        {
            get { return string.Format("{0:yyyy/MM/dd HH:mm:ss:fff}", DateTime.Now); }
        }

        /// <summary> Is logging of command events enabled? </summary>
        public static bool IsLoggingEnabled { get; set; }

        /// <summary>
        /// Log informational or error message
        /// </summary>
        /// <param name="logLevel">Log level</param>
        /// <param name="message">Message to be logged</param>
        public static void Log(LogLevel logLevel, string message)
        {
            if (Logger == null) return;

            // Always log errors regardless of the logging enabled flag
            if (IsLoggingEnabled || logLevel == LogLevel.Error)
                Logger.Log(TimeStamp, logLevel.ToString(), message);
        }
    }
}
